### PR TITLE
Behaviour trees v2

### PR DESCRIPTION
**Pull request by Yisuslalala**

### CODE REVIEW

File: src/RLP_TMR2023/mpu_controller/mpu_controller.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from abc import abstractclassmethod
+
+import time
+from mpu9250_jmdev.registers import *
+from mpu9250_jmdev.mpu_9250 import MPU9250
+
+
+class MpuController(metaclass=Singleton):
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/mpu_controller/mpu_controller.py]: In this file Singleton is not imported

File: src/RLP_TMR2023/mpu_controller/mpu_controller.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from abc import abstractclassmethod
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/mpu_controller/mpu_controller.py]: You should change the name of this directory and file (do not include "controller") as this file is not already a controller, as it is not extending a Singleton or implementing a Mock and factory

File: src/RLP_TMR2023/mpu_controller/mpu_controller.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+while True:
+
+    print("|.....MPU9250 in 0x68 Address.....|")
+    print("Accelerometer", mpu.readAccelerometerMaster())
+    print("Gyroscope", mpu.readGyroscopeMaster())
+    print("Magnetometer", mpu.readMagnetometerMaster())
+    print("Temperature", mpu.readTemperatureMaster())
+    print("\n")
+
+    time.sleep(1)
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/mpu_controller/mpu_controller.py]: Test if this code works

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
+import enum
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py:1]: Make sure to pass flake8 and mypy as first step

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+class CanViewingToBB(py_trees.behaviour.Behaviour):
+    def __init__(self):
+        super().__init__(name="Can distance to BB")
+        #self.blackboard = self.attach_blackboard_client(name=self.name)
+        #self.blackboard.register_key("distance", access=py_trees.common.Access.WRITE)
+        logger.info("Can distance to BB")
+        return py_trees.common.Status.SUCCESS
+
+class Distances(enum.Enum):
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py:30]: I think this enum is good, but it should be defined in another place, mark a TODO comment and is ok

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
+import enum
+import logging
+
+import py_trees.common
+import py_trees.console
+
+logger = logging.getLogger(__name__)
+
+class CanViewingToBB(py_trees.behaviour.Behaviour):
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py:11]: Mark a TODO comment to remove this class as it is only used for testing reasons

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    if distance == Distances.CLOSE:
+        return return_value
+    elif distance == Distances.FAR:
+        return return_value
+    else:
+        return return_value
+
+    return return_value
+
+def create_can_distance() -> py_trees.behaviour.Behaviour:
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: rename this function to specify that it is a subtree

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    can_distance.add_child(CanViewingToBB())
+    return can_distance
+
+def get_actions_subtree() -> py_trees.behaviour.Behaviour:
+    actions = py_trees.composites.Selector(name="Actions", memory=False)
+    actions.add_child(create_can_distance())
+    return actions
+
+# Falta blackboard variable
+def can_distance(distance: Distances) -> py_trees.behaviour.Behaviour:
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: remove this function as I think with the enum it should not be necessary

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    else:
+        return return_value
+
+    return return_value
+
+def create_can_distance() -> py_trees.behaviour.Behaviour:
+    can_viewing_sequence = py_trees.composites.Selector(name="Posibble scenarios", memory=False)
+    can_distance.add_child(py_trees.decorators.EternalGuard(
+        name="Is the can to close?",
+        condition=lambda blackboard: condition_can(blackboard, Distances.CLOSE),
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: remove condition function as use only the lambda

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        # lefts the blackboard variable
+        child=py_trees.behaviours.Success(name="Close")
+    ))
+    can_distance.add_child(py_trees.decorators.EternalGuard(
+        name="Is the can far?",
+        condition=lambda blackboard: condition_can(blackboard, Distances.FAR),
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: same as above

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        #lefts blackboard variable
+        child=py_trees.behaviours.Success(name="Far")
+    ))
+    can_distance.add_child(py_trees.decorators.EternalGuard(
+        name="Is the can in a good distance?",
+        condition=lambda blackboard: condition_can(blackboard, Distances.GOOD),
```
Comment (AgustinZavalaA) [inline src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: same as above

**Review by AgustinZavalaA — CHANGES_REQUESTED** [review]
Please fix the comments added and make sure to pass flake8 and mypy

File: src/RLP_TMR2023/mpu_controller/mpu_controller.py (hunk 3 above)
Comment (Yisuslalala) [reply src/RLP_TMR2023/mpu_controller/mpu_controller.py]: It works boss

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py (hunk 6 above)
Comment (AgustinZavalaA) [reply src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py:11]: do this one

File: src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py (hunk 7 above)
Comment (AgustinZavalaA) [reply src/RLP_TMR2023/prueba_behaviour_trees/can_distance.py]: and this one

**Review by AgustinZavalaA — CHANGES_REQUESTED** [review]

**Review by AgustinZavalaA — APPROVED** [review]

**Pull request merged**